File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        
        def helper(root1, root2):
            if root1 == None and root2 == None:
                return True
            if root1 == None or root2 == None:
                return False
            if root1.val == root2.val:
                return helper(root1.right, root2.right) and helper(root1.left, root2.left)
            return False

        if subRoot == None:
            return True
        if root == None: return False
        
        if helper(root, subRoot): return True
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-7.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int helper(TreeNode *root, int &res){
        if(!root) return 0;
        int left = helper(root->left, res);
        int right = helper(root->right, res);
        if(res < 0 && left == 0 && right == 0){
            res = max(res, root->val);
        }
        int nodeSum = root->val + (left > 0 ? left: 0) + (right > 0 ? right: 0);
        res = max(res, nodeSum);
        root->val = max(0, root->val + max((left > 0? left: 0),  (right > 0 ? right: 0)));
        return root->val;
    }
    int maxPathSum(TreeNode* root) {
        int res = -1001;
        helper(root, res);
        return res;
    }
};
